File: finally_tagless.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE NoImplicitPrelude #-}

import Data.IntMap (insert)
import Prelude (IO, Int, print, (+))

class Expr a

class (Expr a) => Integer a where
  integer :: Int -> a

class (Expr a) => Add a where
  add :: a -> a -> a

instance Expr Int

instance Integer Int where
  integer :: Int -> Int
  integer x = x

instance Add Int where
  add :: Int -> Int -> Int
  add = (+)

lhs :: (Integer a) => a
lhs = integer 1

rhs :: (Integer a) => a
rhs = integer 2

expr :: (Integer a, Add a) => a
expr = add lhs rhs

main :: IO ()
main = do
  print (expr @Int)

File: finally_tagless.ml
module Expr = struct
  module type S = sig
    type a
  end
end

module Integer (S : Expr.S) = struct
  module type S = sig
    val integer : int -> S.a
  end
end

module Add (S : Expr.S) = struct
  module type S = sig
    val add : S.a -> S.a -> S.a
  end
end

module Visitor (S : Expr.S) = struct
  module type S = sig
    include Add(S).S
    include Integer(S).S
  end
end

module Main (S : Expr.S) (V : Visitor(S).S) = struct
  let lhs = V.integer 1
  let rhs = V.integer 2
  let expr = V.add lhs rhs
end

module IMain =
  Main
    (struct
      type a = int
    end)
    (struct
      let integer x = x
      let add = ( + )
    end)

module SMain =
  Main
    (struct
      type a = string
    end)
    (struct
      let integer = string_of_int
      let add = ( ^ )
    end)

let main =
  let module SMain =
    Main
      (struct
        type a = string
      end)
      (struct
        let integer = string_of_int
        let add lhs rhs = "(" ^ lhs ^ " + " ^ rhs ^ ")"
      end)
  in
  let module IMain =
    Main
      (struct
        type a = int
      end)
      (struct
        let integer x = x
        let add = ( + )
      end)
  in
  print_string SMain.expr;
  print_newline ();
  print_int IMain.expr;
  print_newline ();
  ()
